Fix log level filtering in multilog environments

izolyomi opened the pull request:
Fixes issue described in https://github.com/RegenJacob/egui_logger/issues/26

Commit:
Fix log level filtering in multilog environments

RegenJacob commented:
I'll take a look at it when I'm at home.

But currently ci is failing could you maybe take a look at? @izolyomi

izolyomi commented:
I'm sorry for the incomplete PR, I've checked the error, it should work in usual setups. I don't have too much GitHub Ci/Actions experience, but it seems to be caused by some exotic CI configuration disabling `std` features of the `log` crate.
I'm not sure in which environments this crate is supposed to run in, so @RegenJacob please take a look at it yourself as well.

RegenJacob commented:
> I'm sorry for the incomplete PR, I've checked the error, it should work in usual setups. I don't have too much GitHub Ci/Actions experience, but it seems to be caused by some exotic CI configuration disabling `std` features of the `log` crate. I'm not sure in which environments this crate is supposed to run in, so @RegenJacob please take a look at it yourself as well.

Don't stress yourself this is just small hobby project. I'm really glad that people contribute.    

I can confirm that it runs and builds on normal setup. It's just a bit weird that CI fails since I didn't set it up to be a no std build environment (but that's not important right now).

Could you maybe add some documentation and include it in the hello example? (if the PR is ready of course) 

And thanks for your PR!

Commit:
Data structure optimization

izolyomi commented:
Thanks for being open and cooperating, I've just had the chance to have another round with this PR.
I've checked what you meant by the hello example, but I think this PR is waaay simpler than you'd think, it's just bugfix for log level filtering. There's no change in the public API or extra documentation needed and all the examples and tests run without modification. I think it should be considered reasonably ready.

Meanwhile I also took a look at some of your implementation details and added 2 minimal changes for optimization and cleanup. I've just added them in a new commit into this PR. Hm, maybe it would have been luckier to discuss it in a separate PR?
1. Removed the extra hashbrown dependency since the standard HashMap took the hashbrown implementation since a while.
2. Used a ring buffer, i.e. VecDeque instead of Vec to store log records. This way, truncating the log entry list to `max_log_length` using `drain()` does not have to shift all the elements of the vector in memory.

Regarding the CI problem, I'm not sure if I'm even able to see the project settings, how could we solve this together?

RegenJacob commented:
I'm a bit unsure about `VecDeque` as it could hurt performance a bit but this will need some benchmarking. 
Probably better for another PR. 

The rest seems good! 

CI is configured in [.github/workflows/rust.yml](https://github.com/RegenJacob/egui_logger/blob/main/.github/workflows/rust.yml) but it _should_ work as it's just Linux with a normal configuration. 

However if you move the `VecDeque` changes into another PR I'll merge this.

izolyomi commented:
Thanks for the fast reply!

> I'm a bit unsure about `VecDeque` as it could hurt performance a bit

Theoretically, keeping a `Vec` instead of `VecDeque` would hurt performance a lot. Both are backed up by an array, but while Vec prefers keeping elements in a contiguous slice at the cost of poor performance at element removal, VecDeque prefers fast removal of the first elements instead, just like done in `egui_logger`'s `logs.drain(..limit)` in `ui.rs:138` and gets slow at explicit https://doc.rust-lang.org/std/collections/struct.VecDeque.html#method.make_contiguous which is never called in `egui_logger`. See the standard documentation explicitly suggesting to use VecDeque in such cases: https://doc.rust-lang.org/std/vec/struct.Vec.html#method.remove.

However, I haven't done the proper homework to show reasonable benchmarking results to prove these claims, so I don't want to force this change on you.

Regarding the CI, I'll try to somehow fix this in my fork and come back with the results.

RegenJacob commented:
I tried to do some naive Profiling with puffin: 

## 1000 Log entries with `Vec`
![image](https://github.com/user-attachments/assets/4a781a10-83b1-4c9f-b953-2f3c9495bc04)


## 10,000 Log entries with `Vec`
![image](https://github.com/user-attachments/assets/5d5cb5b2-8df7-4076-9592-234bd49c7984)



## 1000 Log entries with `VecDeque`
![image](https://github.com/user-attachments/assets/68b3f7ca-734c-48dc-a84f-5f30b17ca9c9)


## 10,000 Log entries with `VecDeque`
![image](https://github.com/user-attachments/assets/655ceba4-ac02-4c72-8685-9ced20c43763)


So it seems to be about the same for the rendering. Performance issue from #26 still apply.

RegenJacob commented on src/lib.rs:
```diff
@@ -123,14 +132,14 @@ struct Record {
 
 struct Logger {
-    logs: Vec<Record>,
+    logs: VecDeque<Record>,
```
```suggestion
    logs: Vec<Record>,
```

RegenJacob replied:
I think that the Vec isn't important in terms of performance right now. 
It might be interesting to take a look at this when other performance issues have been eliminated.

RegenJacob commented on src/lib.rs:
```diff
@@ -59,19 +67,20 @@ impl Builder {
         if self.enabled(record.metadata()) {
             if let Ok(ref mut logger) = LOGGER.lock() {
-                logger.logs.push(Record {
+                logger.logs.push_back(Record {
```
```suggestion
                logger.logs.push(Record {
```

RegenJacob commented on src/ui.rs:
```diff
@@ -246,7 +246,7 @@ impl LoggerUi {
         let mut logs_displayed: usize = 0;
 
-        let time_padding = logger.logs.last().map_or(0, |record| {
+        let time_padding = logger.logs.back().map_or(0, |record| {
```
```suggestion
        let time_padding = logger.logs.last().map_or(0, |record| {
```

RegenJacob commented on Cargo.toml:
```diff
@@ -16,7 +16,6 @@ include = ["src/*.rs", "Cargo.toml", "LICENSE"]
 egui = "0.31"
 regex = "1.11"
-hashbrown = "0.15"
 
```
```suggestion
[features]
default = ["std"]
std = []

```

RegenJacob reviewed (commented):
I think after postponing the `Vec` Changes and introducing an `std`-feature to the crate to fix CI (and wasm in the future). 

This will be good to go!

RegenJacob commented on src/lib.rs:
```diff
@@ -59,19 +67,20 @@ impl Builder {
     pub fn init(self) -> Result<(), SetLoggerError> {
-        log::set_logger(&EguiLogger).map(|()| log::set_max_level(self.max_level))
+        log::set_max_level(self.max_level);
+        log::set_boxed_logger(Box::new(self.build()))
```
This _should_ fix CI, yet the Issue that this PR fixes will still apply to wasm (or other no_std environments) 
```suggestion
        #[cfg(all(feature = "std", target_has_atomic = "ptr"))]
        log::set_boxed_logger(Box::new(self.build()))?;
        Ok(())
```
It might be smart to add a `std` feature to this crate for future cases.
I think someone _fixed_ wasm before. 
I might take a look at better wasm support in the future though

RegenJacob reviewed (commented):
( oops it seems that i forgot to add those 2 comments?)

izolyomi committed:
Guard init() with cfg annotations. Delay using VecDeque. Follow linter suggestions.

izolyomi committed:
Better logger boxing

izolyomi commented:
Thanks for the benchmark. Ok, I've moved back to `Vec`.
I've also took a turn with `feature`-s, but then inspired from the other PR https://github.com/RegenJacob/egui_logger/pull/31 and better solved the logger lifetime issues without needing `std` features of the `log` crate.
